### PR TITLE
Consolidate: principal-type cache + mgmt-group coverage + Grant/Revoke (stacks on #78)

### PR DESCRIPTION
**Pull request by leet-c1**
## Summary

Three follow-ups stacked on [PR #78](https://github.com/ConductorOne/baton-azure-infrastructure/pull/78):

1. **`perf: cache principal-type lookups`** (5edbf7e) — avoid Graph throttling at customer scale. Lab validation: 76 bindings → 46 unique principals, 30 Graph calls saved.
2. **`feat: management_group resource type + mgmt-group-scope assignments`** (665824c) — adds mgmt-group walk via `armmanagementgroups` SDK + corrects scope attribution. Lab validation: 2 mgmt-group-scope assignments now correctly attributed (was 0).
3. **`feat: Grant/Revoke provisioning`** (3805b85) — wires up `armauthorization.RoleAssignmentsClient.Create/Delete` with idempotency annotations.

Once PR #78 merges, the first commit on this branch (194f198) becomes a no-op and the PR is a clean 3-commit diff.

## Context

Part of the sparse-ACL Azure connector consolidation (see [Session Handoff](https://www.notion.so/3454694ad846815aaa8bc64b69e5ee5d) and PR ductone/c1#16540 for the c1-side classification contract). Once these land, `baton-azure-infrastructure` fully supersedes `baton-azure` functionally: same ScopeBinding emission, richer data surface, typed-SDK auth, plus provisioning.

Still all gated behind `--sync-role-assignments` (default false), so existing CLASSIC customers see no change until they opt in.

## Empirical validation (all against live Azure lab)

| Commit | What I verified | Result |
|---|---|---|
| perf (5edbf7e) | Sync + count unique principals | 76 bindings / 46 unique → 30 Graph calls saved |
| feat mgmt-group (665824c) | Sync + scope-type distribution | `management_group: 2, subscription: 19, resource_group: 56` (correct Tenant-Root Reader attribution; was mis-attributed as sub-scope before) |
| feat Grant/Revoke (3805b85) | Lint / vet / `go test ./...` clean; live Grant/Revoke not tested against lab (SP is read-only) — idempotency covered by unit tests on `isConflict` / `isNotFound` |

## Test plan

- [x] `go fmt ./...` clean
- [x] `go vet ./...` clean
- [x] `go test ./...` clean (full suite)
- [x] `go build ./cmd/baton-azure-infrastructure` clean
- [x] Live sync against the lab tenant (see counts above)
- [ ] Live Grant/Revoke against lab (requires temporary `User Access Administrator` role on runtime SP — TODO in a follow-up session)
- [ ] Reviewer uploads patched c1z into a c1 test tenant; confirms `App.access_model = HYBRID` (requires PR ductone/c1#16540 deployed)

## Known gaps (follow-ups)

- Live Grant/Revoke smoke test (needs lab SP elevation).
- Pre-existing bug flagged in PR #78: `pkg/connector/helper.go:649` maps Graph `#microsoft.graph.group` to `resourceGroupResourceType.Id`. My new `mapGraphPrincipalTypeToBaton` has the correct mapping; did not touch existing helper to keep these PRs narrow. File as a separate issue.
- Sub-resource-scope resource types (KV secrets, Service Bus queues, EventHub, etc.) — NOTES.md item B. Out of scope for this stack.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by leet-c1** [comment]
## Pre-merge review pass — 5 fixes pushed

Did a careful pre-merge review (local code reviewer agent). All 5 findings are now addressed in commits `2644587` and `3c440b8`.

| Concern | Severity | Commit | Fix |
|---|---|---|---|
| Revoke used synced assignment name, but Grant generates fresh UUIDs — wrong-principal revoke possible between Grant and next resync | correctness | `3c440b8` | Query at scope by `(principalId)` via `$filter`, match role UUID locally, DELETE by discovered name. Matches baton-azure pattern. `GrantAlreadyRevoked` when no match. |
| `WithGrantableTo` advertised `groupResourceType` but `batonToAzurePrincipalType` rejects groups — UX leak | UX | `3c440b8` | Dropped `groupResourceType`; comment pins the policy. |
| Stage-1 sub walk aborted whole sync on first 403 | robustness | `3c440b8` | Added `isForbidden(err)` + Warn + `break`, mirroring stage-2 mgmt-group pattern. Mixed-permission tenants now sync partial coverage cleanly. |
| `getPrincipalType` fallback chain broken — returned on first endpoint error instead of trying next | correctness (latent, surfaced at scale) | `2644587` | `continue` on per-endpoint error so `(directoryObjects, users, groups, servicePrincipals)` actually falls through. |
| `principalTypeCache` didn't store misses — every failed lookup re-hit Graph + re-logged Warn per binding | Graph-throttle + log-flood | `3c440b8` | Store `""` on empty/error paths; new `TestPrincipalTypeCache_NegativeHitReturnsEmpty` pins the behavior. |

### Verification

- **Pre-push gauntlet**: `gofmt -l` ✅ · `go vet ./...` ✅ · `golangci-lint run ./...` → 0 issues ✅ · `go test ./...` ✅ · `go build` ✅
- **Lab sync** (PR #79 HEAD `3c440b8`, `--sync-role-assignments`, against lab tenant): 77 `role_assignment` resources emitted with `TRAIT_SCOPE_BINDING`; **77/77 ScopeBindingTrait cross-links resolve** (0 dangling). Scope distribution: 56 resource_group / 19 subscription / 2 management_group.

Ready for a reviewer pass.

**Comment by leet-c1** [comment]
Small bonus fix pushed in `6027d55`: `pkg/connector/role.go:344` had an inverted nil-check (`Properties == nil && RoleDefinitionID != nil`) that panicked in exactly the nil-Properties case it was trying to guard. Fixed to `Properties == nil || RoleDefinitionID == nil`. Flagged by the claude-code review bot as pre-existing but exposed by `00d8d33`'s SDK-v2 sweep. One-line fix; pre-push gauntlet clean.

HEAD: `6027d55`. PR still ready for human review.

**Comment by leet-c1** [comment]
## Proactive pre-merge validation pass — 9 checks, 3 new commits

Ran the full proactive validation gauntlet against lab tenant. Summary of what was done + what was found + what is now fixed.

### Checks performed

| # | Check | Result |
|---|---|---|
| 1 | **c1z pre-fix vs post-fix diff** | Byte-perfect identical except for the intended change: 77/77 role_assignment entitlements dropped `groupResourceType` from GrantableTo. All 5907 entitlement ids + 34889 grant ids match across pre/post. 0 unintended drift. |
| 2 | **Cross-link grant validation** | role_assignment grants: 77/77 principals resolve (0 dangling). Non-RA grants: 926 dangling principal refs at `resource_group` type, all pre-existing `helper.go:649` Entra-group→RG bug (tracked separately). |
| 3 | **Filter-string injection** | Pre-existing risk in Grant/Revoke `$filter` interpolation. **Fixed** in `33f9f2f`: added `isAzureGUID` validator; Grant + Revoke refuse non-GUID principal ids. `TestIsAzureGUID` covers canonical + injection shapes. |
| 4 | **Edge-case unit tests** | **Added** in `0f017ae`: `TestRoleUUIDFromBindingRef`, extended `TestParsePrincipalFromRoleAssignmentResourceID` (multi-`@`, trailing-`@`), extended `TestScopeResourceRefFromAzureScope` (empty, malformed, trailing-slash). |
| 5 | **Entra + infra pair consistency** | Fresh entra sync: 0% drift vs prior (609/1948/143/348/15). Cross-pair user ids: perfect match (609 ∩ 609 = 609). |
| 6 | **Memory buffering risk in List()** | Current: ~1KB/resource × N accumulated in map. Fine at lab scale; at 100K+ assignments would warrant pagination. **Not blocking** — filed as follow-up. |
| 7 | **Diff review of fix commits** | Clean. Minor style note: outer/inner break in Revoke pager loop is slightly redundant but correct. |
| 8 | **Git blame on critical paths** | `role.go` has recurring nil-panic / ID-format history (4 defensive patches incl. tonight's `6027d55`) — area deserves a future rework pass. Not in scope for this PR. |
| 9 | **Live Grant/Revoke E2E via builder** | **New integration test** `role_assignment_live_test.go` (build tag `live`). Caught **4 correctness bugs** the unit tests missed. All fixed in `b60e928`. |

### Bugs caught by #9 (E2E test)

1. **Grant passed composite `<uuid>:<sub>` into ARM path** — produced invalid `.../roleDefinitions/<uuid>:<sub>`, Azure 400. Now uses `roleUUIDFromBindingRef`.
2. **Grant/Revoke passed builder-format scope ids into ARM client** — bare RG name `rg-apps-web-prd` treated as a scope. Added `armScopeFromBindingRef` to reconstruct full ARM paths from `(resourceType, resourceID, subscriptionID)`.
3. **Idempotency classifiers missed gRPC-status errors** — baton-sdk middleware converts Azure 409s into `codes.AlreadyExists`; `isConflict`/`isNotFound`/`isForbidden` only checked `*azcore.ResponseError`. Now check both. `GrantAlreadyExists`/`GrantAlreadyRevoked` annotations were being silently dropped.
4. **$filter syntax rejected by api-version 2022-04-01** — `atScope() and principalId eq 'X'` returns `UnsupportedQuery` on the v2 SDK (v1 accepted combinations). Switched to `atScope()` with local (principal, role) match. Also added 204-with-empty-body detection on Delete for the list-race case.

### Live verification

All 4 E2E phases passed against the lab:
```
Grant#1: success (Azure 201)
Grant#2 (duplicate): GrantAlreadyExists ✓ (Azure 409 → annotation)
Revoke#1: success (Azure 204 with deleted assignment body)
Revoke#2 (missing): GrantAlreadyRevoked ✓ (Azure 204 empty body → annotation)
```

SP was temporarily elevated to `User Access Administrator`, test executed, then de-elevated. Test RG: `rg-apps-web-prd`. Test principal: `sp-aks-placeholder` (lab-intended stale SP).

### Follow-up items (not blocking merge)

- `helper.go:649` Entra-group → RG principal mapping (926 dangling refs flagged by #2)
- `storage_account.go:247` PIM 403 → WARN demotion
- `resource_group_role_assignment.go` dead-code cleanup
- `role.go` comprehensive defensive-decode pass (git blame pattern)
- `List()` pagination-bag refactor for >100K-assignment tenants

### HEAD

`b60e928` — pre-push gauntlet clean (gofmt / vet / golangci-lint 0 issues / all tests pass / build ok).

Recommendation: **ready for human merge review**. The 4 bugs caught in the E2E pass would have been customer-visible failures; we now have both fixes and an integration test that pins the behaviour.

**Comment by leet-c1** [comment]
## Two more fixes pushed in `9e27e61`

### 1. `helper.go:641` — Entra group → Azure RG typo (pre-existing data-quality bug)

Classic copy-paste: `getPrincipalIDResource` routed `#microsoft.graph.group` to `resourceGroupResourceType.Id` instead of `groupResourceType.Id`. The former is the Azure ARM "resource group" concept (bare-name ids like `rg-apps-web-prd`), the latter is Entra/AD directory groups (GUIDs). Every Graph-group principal on a role grant therefore became a dangling reference.

**Surfaced by cross-link validation on this PR:** 926 dangling principal refs in the lab, all mapping to `resource_group` at GUIDs that cannot exist there.

**Callers affected:**
- `role.go:roleBuilder.Grants` — **live code path**, affected every prod sync
- `resource_group_role_assignment.go` — dead/unregistered, would have had the same bug if wired

**Live re-verification post-fix:** 0 dangling principal refs (down from 926).

### 2. `role_assignment.go:List` — large-tenant memory WARN

Asked offline: the 100K figure is **role_assignments**, not users/groups. The two-stage walk (sub + mgmt-group with stage-2 scope overwrite) accumulates every role_assignment resource in memory before flushing to the SDK. At ~1-2 KB per resource:

| Scale | Memory |
|---|---|
| Lab | 77 assignments → ~150 KB — trivial |
| Typical enterprise (~50 subs × 200 assignments) | ~10K → ~15 MB |
| Large enterprise (~500 subs × 500 assignments incl. descendant scopes) | ~250K → ~375 MB |
| Hyperscale (5K+ subs) | 1M+ → 1.5 GB+ (OOM risk) |

User/group syncs are unaffected — they use baton-sdk's pagination tokens cleanly. The issue is isolated to the `role_assignment` builder's stage-1/stage-2 overwrite contract.

Added a WARN log at 50K buffered assignments so large-tenant deployments get operator-visible observability rather than silent OOM. A proper streaming refactor (pagination.Bag with phase state + seen-set) is a tracked follow-up — scope-creep for Apr 27.

### Current HEAD: `9e27e61`

Pre-push gauntlet clean. Live lab sync re-verified: 76/76 role_assignment cross-links, 0 dangling principal refs across all 34,837 grants.

**Comment by leet-c1** [comment]
Follow-up tracked: **[#80](https://github.com/ConductorOne/baton-azure-infrastructure/issues/80) — role_assignment List() pagination refactor for hyperscale tenants.**

Also pushed `1b41a78` with the in-code scale table + refactor-blocker analysis in `role_assignment.go:List` — operators and future maintainers can assess risk at a glance.

Memory profile, for quick reference (per-resource ~1.5 KB):

| Tenant profile | N | RAM |
|---|---:|---:|
| Lab | 77 | ~150 KB |
| SMB | <1K | <2 MB |
| Mid-market | 1-10K | 1.5-15 MB |
| Enterprise | 10-100K | 15-200 MB |
| Large enterprise | 100-500K | 200MB-1GB |
| Hyperscale | 1M+ | 1.5 GB+ |

50K WARN already in place. PR #80 (the issue) has three refactor options evaluated with blockers and a verification checklist.

Not blocking Apr 27 — flag is opt-in, vast majority of customers fall under the threshold, WARN gives observability.

**Comment by leet-c1** [comment]
## Memory concern ✅ FIXED, not just documented — commit `0d1b194`

Refactored `roleAssignmentBuilder.List` from buffered `map[id]*Resource` to streaming `pagination.GenBag[raBagState]`. Closes issue #80.

### What unblocked the refactor

A 5-minute `az rest` check against the lab (api-version 2022-04-01) showed that the comment claiming "stage-1 projects ancestor-inherited scopes to sub level" was **empirically wrong** post–armauthorization v2 SDK bump. The sub walk actually returns mgmt-group-scope-at-tenant-root assignments with their authoritative scope. So the stage-2 overwrite contract was leftover cruft from a pre-SDK-bump era.

Stage-2 is still needed — but only for *intermediate* mgmt groups (between tenant-root and a sub), which don't come through the sub walk at all. That's a small, bounded set that can live in the pagination token's seen-set.

### Before / after

**Before** (commits ≤ `1b41a78`): `byID map[string]*v2.Resource` accumulated every role assignment; memory grew linearly with tenant size. 50K WARN threshold in place as observability stopgap.

**After** (commit `0d1b194`): first call does full mgmt-group enumeration in one go (small set); seed pagination token with pending-sub list + mgmt-group seen-set. Each subsequent call processes exactly one subscription. No unbounded accumulator.

### Memory profile

| | Peak per call | Between calls (token) |
|---|---|---|
| Typical enterprise | 150-750 KB (one sub, 100-500 assignments) | ~10-50 KB (pending subs) |
| Hyperscale sub (10K assignments) | ~15 MB | Bounded by N subs, not total assignments |
| Total-tenant-assignment count | N/A (doesn't accumulate) | ~216 KB worst mainstream (5K subs + 1K mgmt-scope) |

### Verification

- Output **byte-identical** to pre-refactor baseline: 76 role_assignments, 76/76 cross-links resolve, scope distribution 2 mgmt-group / 19 sub / 55 RG, 5906 entitlements, 34837 grants — all match exactly.
- Live Grant/Revoke integration test (`-tags=live`) **still passes**: Grant#1 success, Grant#2 GrantAlreadyExists, Revoke#1 success, Revoke#2 GrantAlreadyRevoked.
- Pre-push gauntlet clean.
- The misleading "projected scope" comment is gone; new docstring explains the actual SDK v2 behaviour and what stage-2 is still for.

### Lesson

SDK version bumps invalidate comment-level assumptions that are not re-verified against behavior. This class of drift is invisible until a scale concern forces you to re-read the comment and reality no longer matches. Catching this one saved us ~1.5 GB at hyperscale.

HEAD: `0d1b194`. Issue #80 closed.

File: pkg/connector/role_assignment.go [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+// azure-sdk-for-go typed armauthorization client rather than hand-rolled REST.
+type roleAssignmentBuilder struct {
+	conn *Connector
+
+	// principalTypeCache memoizes getPrincipalType(ctx, conn, principalID) results
+	// across calls. A single principal typically holds many role assignments, so
+	// without this cache Grants() would hit Microsoft Graph's directoryObjects
+	// endpoint once per binding — throttling risk at customer scale. sync.Map
+	// because Grants may be invoked concurrently per resource.
+	principalTypeCache sync.Map // map[principalID string] -> principalType string
```
Comment (kans) [inline pkg/connector/role_assignment.go:50]: 1. How much data goes in here at scale? 
2. If we run this on lambdas, this cache will be dropped and recreated per ResourceSyncer method call.

**Comment by leet-c1** [comment]
Superseded/abandoned.
